--- xpdf/CharCodeToUnicode.cc.orig	2012-06-11 09:11:40.000000000 +0200
+++ xpdf/CharCodeToUnicode.cc	2012-06-11 09:48:58.000000000 +0200
@@ -395,7 +395,13 @@
     if (code >= mapLen) {
       mapLen = (code + 256) & ~255;
     }
+    Unicode *oldmap = map;
     map = (Unicode *)greallocn(map, mapLen, sizeof(Unicode));
+    if (!map) {
+      /* we sometimes get overflows for files trying to use 0xffffffff as charcode */
+      map = oldmap;
+      return;
+    }
     for (i = oldLen; i < mapLen; ++i) {
       map[i] = 0;
     }
